[CONTINT-3437][language detection] decouple handler and patcher via workloadmeta

adel121 opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->

### ATTENTION 
This PR is better reviewed commit-by-commit
### What does this PR do?

This PR decouples the language detection handler from the language detection patcher. The handler will now push detected languages into workloadmeta, and the patcher will subscribe to workloadmeta and patch pod owners (i.e deployments, statefulsets, etc.) when processing events.
<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
Language detection had 3 main components:
- client: runs on each node (within the agent), receives information about detected process languages, aggregates them by pod, and send them to the handler/server
- handler/server: runs in the cluster agent. receives and handles language detection requests from the client. It is responsible to communicate with the kubernetes API Server in order to update/patch deployments language annotations.
- admission controller: handles pod mutating requests and injects required language libraries into the applicative pods.

Initially, the handler was performing the patching operations directly upon receiving requests from the language detection client. This can lead to performance issues when the number of requests received increases a lot, which will happen in case of large deployments for example (thousands of pods, where each pod runs several containers/processes). 

This PR extracts the patching logic into a separate component (the language detection patcher), and decouples the handler from the patcher via workloadmeta. 

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->
- In order to have a clean implementation, some refactors done:
  - Centralising all language-detection specific structs and methods in one package `pkg/languagedetection/util`
  - Refactoring deployment entities in workloadmeta to re-use the language detection package to define languages

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->


### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

To QA these changes, deploy the agent and cluster agent with the following:
- language detection enabled on both the agent and cluster agent
- process agent enabled and process collection enabled
- sufficient rbac for the cluster agent to patch deployments
- admission controller enabled and `mutateUnlabelled` set to `true`

Then create a deployment with some containers running dummy processes with supported languages (e.g. ruby, python, java, go). 

Check for the following:
- The deployment is patched with the correct languages
- The deployment content of workloadmeta is correct (shows the detected and injectable languages)

Here are the steps:

1- Deploy the agent and cluster agent using helm:
```
datadog:
  apiKeyExistingSecret: datadog-secret
  appKeyExistingSecret: datadog-secret
  kubelet:
    tlsVerify: false
  processAgent:
    enabled: true
    processCollection: true
  env:
    - name: DD_LANGUAGE_DETECTION_ENABLED
      value: "true"

agents:
  telemetry: enabled

clusterAgent:
  enabled: true
  replicas: 1
  admissionController:
    enabled: true
    mutateUnlabelled: true
  env:
    - name: DD_LANGUAGE_DETECTION_ENABLED
      value: "true"

```

2- Allow the cluster agent to patch deployment by updating the clusterrole. Use the following command: `kubectl edit clusterrole datadog-agent-cluster-agent`, then add the `patch` action to the list of verbs for any `apiGroup` containing deployment as a resource.

3- create a deployment with 2 containers (the example below uses an app that has 2 containers, one running `python` and one running ` ruby`):
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dummy-user-deployment
  labels:
    app: user-app
spec:
  replicas: 3
  selector:
    matchLabels:
      app: user-app
  template:
    metadata:
      labels:
        app: user-app
    spec:
      containers:
        - name: dummy-ruby-container
          image: ruby:2.7-slim # You can replace this with the desired Ruby base image
          command: ["ruby", "-e", "loop { sleep 1000 }"] # Ruby script to sleep forever
        - name: python-process-container
          image: python:3.7 # Replace with your Python image
          command: ["python3", "-u", "your-python-script.py"]
          volumeMounts:
            - name: python-script-volume
              mountPath: /your-python-script.py
              subPath: your-python-script.py
      volumes:
        - name: python-script-volume
          configMap:
            name: python-script-configmap
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: python-script-configmap
data:
  your-python-script.py: |
    while True:
      pass

```

4- Wait a few seconds, then inspect the content of workloadmeta in the cluster agent. It should show the detected languages for the deployments as such:
```
=== Entity kubernetes_deployment sources(merged):[kubeapiserver language_detection_server] id: default/dummy-user-deployment ===
----------- Entity ID -----------
Kind: kubernetes_deployment ID: default/dummy-user-deployment

----------- Unified Service Tagging -----------
Env : 
Service : 
Version : 
----------- Injectable Languages -----------
Container dummy-ruby-container=>[ruby]
Container python-process-container=>[python]
----------- Detected Languages -----------
Container python-process-container=>[python]
Container dummy-ruby-container=>[ruby]
===
```

5- Verify that language annotations were added to the deployment:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    internal.dd.datadoghq.com/dummy-ruby-container.detected_langs: ruby
    internal.dd.datadoghq.com/python-process-container.detected_langs: python
```

6- Now update the deployment by removing the python container and adding another ruby container.
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dummy-user-deployment
  labels:
    app: user-app
spec:
  replicas: 3
  selector:
    matchLabels:
      app: user-app
  template:
    metadata:
      labels:
        app: user-app
    spec:
      containers:
        - name: dummy-ruby-container-extra
          image: ruby:2.7-slim # You can replace this with the desired Ruby base image
          command: ["ruby", "-e", "loop { sleep 1000 }"] # Ruby script to sleep forever
        - name: dummy-ruby-container
          image: ruby:2.7-slim # You can replace this with the desired Ruby base image
          command: ["ruby", "-e", "loop { sleep 1000 }"] # Ruby script to sleep forever
```

7-  Wait a few seconds, then inspect the content of workloadmeta in the cluster agent. It should show the new ruby container with its detected language. Python should not be removed since the cleanup mechanism is not yet implemented
```
=== Entity kubernetes_deployment sources(merged):[kubeapiserver language_detection_server] id: default/dummy-user-deployment ===
----------- Entity ID -----------
Kind: kubernetes_deployment ID: default/dummy-user-deployment

----------- Unified Service Tagging -----------
Env : 
Service : 
Version : 
----------- Injectable Languages -----------
Container python-process-container=>[python]
Container dummy-ruby-container-extra=>[ruby]
Container dummy-ruby-container=>[ruby]
----------- Detected Languages -----------
Container python-process-container=>[python]
Container dummy-ruby-container-extra=>[ruby]
Container dummy-ruby-container=>[ruby]
===
```

8- Verify the language annotations are as expected
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
    internal.dd.datadoghq.com/dummy-ruby-container-extra.detected_langs: ruby
    internal.dd.datadoghq.com/dummy-ruby-container.detected_langs: ruby
    internal.dd.datadoghq.com/python-process-container.detected_langs: python
```

9- Finally, verify that libraries were injected as init containers on the user applicative pods (note that `python` library will still be injected because language detection injects the languages that were detected in the previous deployment rollout, you can try to scale the deployment down to 0 and then scale it up, only python library should be injected):
```
initContainers:
  - command:
    - sh
    - copy-lib.sh
    - /datadog-lib
    env:
    - name: DD_INSTRUMENTATION_INSTALL_ID
    - name: DD_INSTRUMENTATION_INSTALL_TIME
      value: "1707397778"
    - name: DD_ENTITY_ID
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.uid
    - name: DD_DOGSTATSD_URL
      value: unix:///var/run/datadog/dsd.socket
    - name: DD_TRACE_AGENT_URL
      value: unix:///var/run/datadog/apm.socket
    image: gcr.io/datadoghq/dd-lib-python-init:latest
    imagePullPolicy: Always
    name: datadog-lib-python-init
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /datadog-lib
      name: datadog-auto-instrumentation
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-vtt9j
      readOnly: true
  - command:
    - sh
    - copy-lib.sh
    - /datadog-lib
    env:
    - name: DD_INSTRUMENTATION_INSTALL_ID
    - name: DD_INSTRUMENTATION_INSTALL_TIME
      value: "1707397778"
    - name: DD_ENTITY_ID
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.uid
    - name: DD_DOGSTATSD_URL
      value: unix:///var/run/datadog/dsd.socket
    - name: DD_TRACE_AGENT_URL
      value: unix:///var/run/datadog/apm.socket
    image: gcr.io/datadoghq/dd-lib-ruby-init:latest
    imagePullPolicy: Always
    name: datadog-lib-ruby-init
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /datadog-lib
      name: datadog-auto-instrumentation
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-vtt9j
      readOnly: true
```
### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

adel121 committed:
Refactor language detection util package

adel121 committed:
re-use language detection util package to store languages as sets in deployment entities

adel121 committed:
update language detection client to use language detection util package

adel121 committed:
refactor language detection handler to push detected languages in workloadmeta instead of invoking the patcher

adel121 committed:
update admission controller to use the new data structure of languages in workloadmeta

adel121 committed:
refactor the patcher to subscribe to workloameta and handle languagedetection events

sgnn7 reviewed (approved):
OK for ASC owned `comp/core/hostname/hostnameinterface/go.sum`

adel121 commented on pkg/languagedetection/util/owners.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// SupportedBaseOwners is the set of owner resource types supported by language detection
+// Currently only deployments are supported
```
It is already mentioned in the comment of the `GetNamespacedBaseOwnerReference` method.

To add, the first part just defines the kinds, and then we have a set of supported kinds.

But in any case, better clarity is good, I will add a note in the comment just above `SupportedBaseOwners` 

thanks 👍

adel121 commented on pkg/languagedetection/util/containerlanguages.go:
```diff
@@ -0,0 +1,132 @@
+// EqualTo checks if current ContainersLanguages object has identical content
+// in comparison another ContainersLanguages
+func (c ContainersLanguages) EqualTo(other ContainersLanguages) bool {
+	if other == nil {
```
I think if `c` is `nil`, we can't invoke this method in the first place. Isn't it?

adel121 commented on cmd/cluster-agent/subcommands/start/command.go:
```diff
@@ -345,6 +346,12 @@ func start(log log.Component, config config.Component, taggerComp tagger.Compone
 		}()
 	}
 
+	if pkgconfig.Datadog.GetBool("language_detection.enabled") {
```
I think we should not rely here on admission controller, because in that case we will not be able to evaluate/test language detection while also disabling the admission controller.

I think this capability is important, at least internally, to be able to see what languages are being detected, how annotations are being patched and also evaluate the impact on load on k8s api server. All this should be done while admission controller is not injecting libraries of detected languages.

WDYT?

adel121 committed:
PR review

adel121 commented:
Thanks @davidor for all your comments and suggestions.

I have taken all of them into consideration, and here is what is included in the most recent commit that I have added:
- Responding to all your comments and suggestions
- Fixing an aliasing issue with workloadmeta. I was storing the ContainersLanguages (which is a map) in workloadmeta. This is stored as a reference by default, leading to aliasing (i.e. the source map and the map stored in workloadmeta are pointing to the same value in memory, thus causing a lot of issues in workloadmeta). I added a `DeepCopy` method to fix this.
- Fixed `ToProto` method of `ContainerLanguages` to distinguish between init and non-init containers.
- Added a new unit test for the PLD handler to ensure the behaviour is as expected in case of successive language detection requests.
- Simplified the singleton pattern used in the patcher component.
- Fixed minor typos

Thanks again

adel121 committed:
add references in docs

davidor commented on pkg/clusteragent/languagedetection/README.md:
```diff
@@ -0,0 +1,14 @@
+# Language Detection Patcher
```
This contains the same description as the `doc.go` file so I think one should be removed.

adel121 committed:
Merge branch 'main' into adelhajhassan/pld_decouple_handler_and_patcher

adel121 committed:
Merge branch 'main' into adelhajhassan/pld_decouple_handler_and_patcher

adel121 committed:
Merge branch 'main' into adelhajhassan/pld_decouple_handler_and_patcher

adel121 committed:
Merge branch 'main' into adelhajhassan/pld_decouple_handler_and_patcher

adel121 commented:
/merge

adel121 committed:
Merge branch 'main' into adelhajhassan/pld_decouple_handler_and_patcher

adel121 committed:
Merge branch 'main' into adelhajhassan/pld_decouple_handler_and_patcher

adel121 committed:
remove maps package and use DeepCopy instead

GustavoCaso commented on cmd/cluster-agent/api/v1/languagedetection/telemetry.go:
```diff
@@ -0,0 +1,36 @@
+
+var (
+	// OkResponses tracks the number the request was processed successfully
+	OkResponses = telemetry.NewCounterWithOpts(
+		subsystem,
+		"ok_response",
+		[]string{},
+		"Tracks the number of times request processing succeeds",
+		commonOpts,
+	)
+
+	// ErrorResponses tracks the number of times request processsing fails
+	ErrorResponses = telemetry.NewCounterWithOpts(
+		subsystem,
+		"fail_response",
+		[]string{},
+		"Tracks the number of times request processing fails",
+		commonOpts,
+	)
```
Is there a way we could avoid creating Global variables for the `languagedetection` package? Are this variables used outside of the `languagedetection` package? If not we might be able to instantiate them when the handler `InstallLanguageDetectionEndpoints` gets called.

adel121 replied:
Hey @GustavoCaso 

Sorry didn't pay attention to your comment before merging.

Will take your suggestion into consideration in a separate PR

Thanks 🙇‍♂️
